core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -2956,7 +2963,12 @@ public static Object structAccess(Object structObject, int index, String fieldNa
     } else {
       Class<?> beanClass = structObject.getClass();
       try {
-        Field structField = beanClass.getDeclaredField(fieldName);
+        Field structField;
+        if (fieldName == null) {
+          structField = beanClass.getDeclaredFields()[index];
```

Comment:
@asolimando , @zabetak , `getDeclaredFields()` is JVM-specific, and the order of the fields is unspecified :(

The exception `RESOURCE.failedToAccessField(fieldName` below should have been updated to print `index` in case of failure.

Reply:
Thanks for taking a look @vlsi . 

When we create the model/types we rely more or less on the same methods (e.g., https://github.com/apache/calcite/blob/3ccad7b2bcc962d48c1dee1c9604fd085d82c608/core/src/main/java/org/apache/calcite/jdbc/JavaTypeFactoryImpl.java#L75) so in many cases the code should work fine I think. I don't know how we can handle this better with the way the code is right now. Do you have a better idea?

It's true that the exception should be updated. I will take care of it but let's see if there are other things to do at the same time.

Reply:
The issue is known as https://issues.apache.org/jira/browse/CALCITE-2489

I suggest we refrain from adding undefined behavior like `getDeclaredFields()[index]`, especially taking into account JVM diversity we have nowadays.


The problem here is the implementation in the current PR might result in silent data corruption. For instance, if the input objects have different types (e.g. different subtypes of the same base type), then the order of the fields might vary across subtypes :-/

So I would suggest:
a) Use `relType.getFields()[index]` if it is possible (note, I mean struct field index rather reflection field index)
b) Hide the feature under a disabled-by-default feature flag. Then users won't accidentally corrupt data "thanks" to an experimental feature.

Reply:
At first glance, option a) does not seem very easy to implement. 

Regarding option b) do you expect an exception at runtime when we reach this part of the code and the feature is disabled or something at a higher layer?

Reply:
> b)

For instance, we could keep the old behavior.
I guess the previous behavior was "throw NPE since fieldName is null" which is not user-friendly.
So the fix might be to throw `IllegalArgumentException("fieldName is unknown, fieldIndex is ..., you might add calcite.experimental.allow.field.index.access=true to allow index-based field access")` or something like that.

I think a feature-flag here does not incur high maintenance cost, however, I think we should refrain from making undefined behavior-like features activated by default.

Reply:
@vlsi Sounds good! 

@asolimando do you wanna take this follow up or should I put it on me?

Reply:
@vlsi thanks for taking a look, I agree on the whole line.

@zabetak thanks for asking, I can take care of implementing (b) for a start, I just have few additional questions:

1. Do we need to file a new Jira ticket for this? If not, which are the conventions for PR name and branch name in such cases (PR closed, a fix is needed)?

 2. What happens to the unit-tests relying on the feature? (I had a look at `CalciteSystemProperty.java` but there are no tests relying on enabling properties). 
If they were JUnit tests, I'd conditionally enable them based on the `calcite.experimental.allow.field.index.access` property, but I am not sure how to achieve that for quidem tests.

In addition, test `testItemOp()` in `core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java` has been extended to cover the new functionality and should not be disabled entirely IMO. Would wrapping the part relying on the extended feature with a:
```java
if (CalciteSystemProperty.FIELD_INDEX_ACCESS.value()) {
...
}
```
be OK?

Reply:
I have figured it out what was not clear at first and I have opened [PR-2267](https://github.com/apache/calcite/pull/2267).

Can I ask you to review it if you have time?